File: apps/nucleusd/src/query/typed_response/selected_task_product_aggregate.rs
use nucleus_server::ControlSelectedTaskProductAggregateDto;

pub(crate) fn selected_task_product_aggregate_response_lines(
    label: &str,
    aggregate: ControlSelectedTaskProductAggregateDto,
) -> Vec<String> {
    vec![
        format!("domain={label}"),
        format!("aggregate_id={}", aggregate.aggregate_id),
        format!("project_id={}", aggregate.project_id),
        format!("task_id={}", aggregate.task_id),
        format!(
            "task title={} activity={} assignment={} action_type={}",
            aggregate.identity.title.as_deref().unwrap_or("none"),
            aggregate.identity.activity.as_deref().unwrap_or("none"),
            aggregate.identity.assignment.as_deref().unwrap_or("none"),
            aggregate.identity.action_type.as_deref().unwrap_or("none")
        ),
        format!(
            "next phase={} action={} next_ref={} blocked_reason={}",
            aggregate.workflow.phase,
            aggregate.workflow.primary_next_action,
            aggregate.workflow.next_ref.as_deref().unwrap_or("none"),
            aggregate
                .workflow
                .blocked_reason
                .as_deref()
                .unwrap_or("none")
        ),
        format!(
            "readiness allowed_actions={} blockers={} unavailable_actions={}",
            aggregate.readiness.allowed_action_count,
            aggregate.readiness.blockers.len(),
            aggregate.readiness.unavailable_actions.len()
        ),
        format!(
            "command_previews admitted={} refused={} total={}",
            aggregate.command_previews.admitted_count,
            aggregate.command_previews.refused_count,
            aggregate.command_previews.previews.len()
        ),
        format!(
            "work_evidence work_items={} active={} completed={} evidence_refs={} timeline_refs={}",
            aggregate.work_evidence.work_item_refs.len(),
            aggregate.work_evidence.active_work_item_count,
            aggregate.work_evidence.completed_work_item_count,
            aggregate.work_evidence.evidence_refs.len(),
            aggregate.work_evidence.timeline_refs.len()
        ),
        format!(
            "review state={} route_status={} primary_route={} decision_available={} evidence_refs={}",
            aggregate.review.state.as_deref().unwrap_or("none"),
            aggregate.review.route_status.as_deref().unwrap_or("none"),
            aggregate.review.primary_route.as_deref().unwrap_or("none"),
            aggregate.review.decision_available,
            aggregate.review.evidence_refs.len()
        ),
        format!(
            "rework status={} reviewed_work_item_refs={} reviewed_evidence_refs={} refusal={}",
            aggregate.rework.status.as_deref().unwrap_or("none"),
            aggregate.rework.reviewed_work_item_refs.len(),
            aggregate.rework.reviewed_evidence_refs.len(),
            aggregate.rework.refusal_reason.as_deref().unwrap_or("none")
        ),
        format!(
            "completion status={} command_available={} evidence_refs={} refusal={}",
            aggregate.completion.status.as_deref().unwrap_or("none"),
            aggregate.completion.command_available,
            aggregate.completion.evidence_refs.len(),
            aggregate.completion.refusal_reason.as_deref().unwrap_or("none")
        ),
        format!(
            "scm_handoff state={} next_category={} target_shape={} evidence_refs={} gaps={}",
            aggregate.scm_handoff.state.as_deref().unwrap_or("none"),
            aggregate
                .scm_handoff
                .next_category
                .as_deref()
                .unwrap_or("none"),
            aggregate
                .scm_handoff
                .target_shape
                .as_deref()
                .unwrap_or("none"),
            aggregate.scm_handoff.evidence_refs.len(),
            aggregate.scm_handoff.gap_count
        ),
        format!(
            "source_health sources={} missing={} partial={} gaps={}",
            aggregate.source_health.sources.len(),
            aggregate.source_health.missing_count,
            aggregate.source_health.partial_count,
            aggregate.gaps.len()
        ),
        format!(
            "no_effects task_mutation={} provider_execution={} provider_write={} scm_or_forge_mutation={} accepted_memory_apply={} planning_apply={} projection_write={} agent_scheduling={} ui_effect={}",
            aggregate.no_effects.task_mutation_performed,
            aggregate.no_effects.provider_execution_performed,
            aggregate.no_effects.provider_write_performed,
            aggregate.no_effects.scm_or_forge_mutation_performed,
            aggregate.no_effects.accepted_memory_apply_performed,
            aggregate.no_effects.planning_apply_performed,
            aggregate.no_effects.projection_write_performed,
            aggregate.no_effects.agent_scheduling_performed,
            aggregate.no_effects.ui_effect_performed
        ),
        "mode=read_only_product_aggregate".to_owned(),
        "client_can_mutate=false".to_owned(),
        "provider_execution_available=false".to_owned(),
        "scm_or_forge_execution_available=false".to_owned(),
        "proof_payload_dump=false".to_owned(),
        "payloads_exposed=false".to_owned(),
    ]
}

File: apps/desktop/src/lib/control/selectedTaskProductAggregate.ts
import { parseSingleRecordResponse, type QueryFallback } from "./singleRecordResponse";
import {
  buildControlQueryEnvelope,
  type ControlRequestEnvelopeDto,
  type ControlResponseEnvelopeDto,
} from "./envelopes";
import type { ControlTaskWorkflowNoEffectsDto } from "./taskWorkflow";
import { CONTROL_CLIENT_ID } from "./types";

export type ControlSelectedTaskProductIdentityDto = {
  title: string | null;
  activity: string | null;
  assignment: string | null;
  action_type: string | null;
  expected_revision: string | null;
};

export type ControlSelectedTaskProductWorkflowDto = {
  primary_next_action: string;
  reason: string;
  phase: string;
  next_ref: string | null;
  blocked_reason: string | null;
};

export type ControlSelectedTaskProductBlockerDto = {
  family: string;
  reason: string;
  evidence_refs: string[];
};

export type ControlSelectedTaskProductUnavailableActionDto = {
  family: string;
  status: string;
  reason: string;
};

export type ControlSelectedTaskProductReadinessDto = {
  blockers: ControlSelectedTaskProductBlockerDto[];
  unavailable_actions: ControlSelectedTaskProductUnavailableActionDto[];
  allowed_action_count: number;
};

export type ControlSelectedTaskProductCommandPreviewDto = {
  family: string;
  status: string;
  command_available: boolean;
  refusal_reason: string | null;
  evidence_refs: string[];
};

export type ControlSelectedTaskProductCommandPreviewsDto = {
  admitted_count: number;
  refused_count: number;
  previews: ControlSelectedTaskProductCommandPreviewDto[];
};

export type ControlSelectedTaskProductWorkEvidenceDto = {
  work_item_refs: string[];
  active_work_item_count: number;
  completed_work_item_count: number;
  evidence_refs: string[];
  timeline_refs: string[];
};

export type ControlSelectedTaskProductReviewDto = {
  state: string | null;
  next_category: string | null;
  route_status: string | null;
  primary_route: string | null;
  decision_ref: string | null;
  decision_available: boolean;
  blocker_reasons: string[];
  evidence_refs: string[];
};

export type ControlSelectedTaskProductReworkDto = {
  status: string | null;
  summary: string | null;
  refusal_reason: string | null;
  reviewed_work_item_refs: string[];
  reviewed_evidence_refs: string[];
};

export type ControlSelectedTaskProductCompletionDto = {
  status: string | null;
  command_available: boolean;
  refusal_reason: string | null;
  evidence_refs: string[];
};

export type ControlSelectedTaskProductScmHandoffDto = {
  state: string | null;
  next_category: string | null;
  target_shape: string | null;
  blocker_refs: string[];
  evidence_refs: string[];
  gap_count: number;
};

export type ControlSelectedTaskProductSourceStatusDto = {
  source: string;
  state: string;
  reason: string | null;
};

export type ControlSelectedTaskProductSourceHealthDto = {
  sources: ControlSelectedTaskProductSourceStatusDto[];
  missing_count: number;
  partial_count: number;
};

export type ControlSelectedTaskProductGapDto = {
  source: string;
  reason: string;
};

export type ControlSelectedTaskProductAggregateDto = {
  aggregate_id: string;
  project_id: string;
  task_id: string;
  identity: ControlSelectedTaskProductIdentityDto;
  workflow: ControlSelectedTaskProductWorkflowDto;
  readiness: ControlSelectedTaskProductReadinessDto;
  command_previews: ControlSelectedTaskProductCommandPreviewsDto;
  work_evidence: ControlSelectedTaskProductWorkEvidenceDto;
  review: ControlSelectedTaskProductReviewDto;
  rework: ControlSelectedTaskProductReworkDto;
  completion: ControlSelectedTaskProductCompletionDto;
  scm_handoff: ControlSelectedTaskProductScmHandoffDto;
  source_health: ControlSelectedTaskProductSourceHealthDto;
  gaps: ControlSelectedTaskProductGapDto[];
  no_effects: ControlTaskWorkflowNoEffectsDto;
};

export type SelectedTaskProductAggregateQueryResult =
  | {
      state: "record";
      aggregate: ControlSelectedTaskProductAggregateDto;
    }
  | QueryFallback;

export function buildSelectedTaskProductAggregateQuery(
  projectId: string,
  taskId: string,
  expectedRevision: string | null,
): ControlRequestEnvelopeDto {
  return buildControlQueryEnvelope({
    kind: "selected_task_product_aggregate",
    query_id: "",
    action: "aggregate",
    project_id: projectId,
    task_id: taskId,
    expected_revision: expectedRevision,
    operator_ref: CONTROL_CLIENT_ID,
  });
}

export function selectedTaskProductAggregateFromResponse(
  response: ControlResponseEnvelopeDto,
): SelectedTaskProductAggregateQueryResult {
  return parseSingleRecordResponse(response, "selected_task_product_aggregate", "selected task product aggregate", (body) => ({
    state: "record" as const,
    aggregate: body.aggregate,
  }));
}
